source/extensions/filters/http/lua/lua_filter.h

```diff
@@ -21,23 +22,27 @@ constexpr char GLOBAL_SCRIPT_NAME[] = "GLOBAL";
 
 class PerLuaCodeSetup : Logger::Loggable<Logger::Id::lua> {
 public:
-  PerLuaCodeSetup(const std::string& lua_code, ThreadLocal::SlotAllocator& tls);
+  PerLuaCodeSetup(const std::string& lua_code, ThreadLocal::SlotAllocator& tls,
+                  Event::Dispatcher& dispatcher);
+  ~PerLuaCodeSetup() { dispatcher_.deferredDelete(std::move(lua_state_)); }
```

Comment:
What was the reason you added deferred delete here? I think this part is a little tricky unfortunately.

Basically, I think it's possible for the route table to get removed from the main thread but still be stored by an active request. Thus, the request may hold that last reference to the route table and then when it's over it will destroy the route table and the per-lua code. I think this means that this may happen cross thread. Is deferred delete safe cross thread? (It may be I would need to check.) Anyway this is the part that needs very careful thought. Can you describe what you did here and add more comments so we can discuss?

Reply:
Main thread dispatcher
https://github.com/envoyproxy/envoy/blob/c9c4709c844b90b9bb2935d784a428d667c9df7d/include/envoy/server/factory_context.h#L50-L54

PerLuaCodeSetup will keep a reference of main thread dispatcher
https://github.com/wbpcode/envoy/blob/ba370561fae8c7392c272a1fe60d6ce996c1f505/source/extensions/filters/http/lua/lua_filter.cc#L635-L637



This 'dispatcher_' belongs to the main thread. And add this deferred delete will move TheadLocalState(lua_state_) and destroy it in the main thread.

In this case, no matter on which thread the route table is destroyed, we can guarantee that TheadLocalState(lua_state_) will only be destroyed in the main thread. 

In this way, we can ensure that the state of TLS will not be modified on the worker thread.

Reply:
I rechecked the source code and found that `deferredDelete` itself is not thread-safe, and it cannot safe cross thread. But in the release version, this problem is not easy to appear.

Reply:
It's possible this will work but the threading would need to be analyzed. Do you want to try this out and we can take a look?

Reply:
Ok. I will update it this weekend. My own branch also needs to verify this function in the production environment immediately.